src/couch/src/couch_db_split.erl

```diff
@@ -0,0 +1,491 @@
+    end.
+
+
+purge_cb({_PSeq, _UUID, Id, _Revs} = PI, #state{targets = Targets} = State) ->
+    {Key, Target} = pick_target(Id, State, Targets),
+    MaxBuffer = State#state.max_buffer_size,
+    Target1 = acc_and_flush(PI, Target, MaxBuffer, fun commit_purge_infos/1),
+    {ok, State#state{targets = Targets#{Key => Target1}}}.
+
+
+commit_purge_infos(#target{buffer = [], db = Db} = Target) ->
+    Target#target{db = Db};
+
+commit_purge_infos(#target{buffer = PIs0, db = Db} = Target) ->
+    PIs = lists:reverse(PIs0),
+    {ok, Db1} = couch_db_engine:copy_purge_infos(Db, PIs),
+    {ok, Db2} = couch_db_engine:commit_data(Db1),
+    Target#target{buffer = [], buffer_size = 0, db = Db2}.
+
+
+copy_docs(#state{source_db = Db} = State) ->
+    {ok, NewState} = couch_db:fold_changes(Db, 0, fun changes_cb/2, State),
+    CommitTargets = maps:map(fun(_, #target{} = T) ->
+        commit_docs(T)
+    end, NewState#state.targets),
+    NewState#state{targets = CommitTargets}.
+
+
+% Backwards compatibility clause. Seq trees used to hold #doc_infos at one time
+changes_cb(#doc_info{id = Id}, #state{source_db = Db} = State) ->
+    [FDI = #full_doc_info{}] = couch_db_engine:open_docs(Db, [Id]),
+    changes_cb(FDI, State);
+
+changes_cb(#full_doc_info{id = Id} = FDI, #state{} = State) ->
+    #state{source_db = SourceDb, targets = Targets} = State,
+    {Key, Target} = pick_target(Id, State, Targets),
+    FDI1 = process_fdi(FDI, SourceDb, Target#target.db),
+    MaxBuffer = State#state.max_buffer_size,
+    Target1 = acc_and_flush(FDI1, Target, MaxBuffer, fun commit_docs/1),
+    {ok, State#state{targets = Targets#{Key => Target1}}}.
+
+
+commit_docs(#target{buffer = [], db = Db} = Target) ->
+    Target#target{db = Db};
+
+commit_docs(#target{buffer = FDIs, db = Db} = Target) ->
+    Pairs = [{not_found, FDI} || FDI <- lists:reverse(FDIs)],
+    {ok, Db1} = couch_db_engine:write_doc_infos(Db, Pairs, []),
+    {ok, Db2} = couch_db_engine:commit_data(Db1),
+    Target#target{buffer = [], buffer_size = 0, db = Db2}.
+
+
+process_fdi(FDI, SourceDb, TargetDb) ->
+    #full_doc_info{id = Id, rev_tree = RTree} = FDI,
+    Acc = #racc{id = Id, source_db = SourceDb, target_db = TargetDb},
+    {NewRTree, NewAcc} = couch_key_tree:mapfold(fun revtree_cb/4, Acc, RTree),
+    {Active, External} = total_sizes(NewAcc),
+    FDI#full_doc_info{
+        rev_tree = NewRTree,
+        sizes = #size_info{active = Active, external = External}
+    }.
+
+
+revtree_cb(_Rev, _Leaf, branch, Acc) ->
+    {[], Acc};
+
+revtree_cb({Pos, RevId}, Leaf, leaf, Acc) ->
+    #racc{id = Id, source_db = SourceDb, target_db = TargetDb} = Acc,
+    #leaf{deleted = Deleted, ptr = Ptr} = Leaf,
+    Doc0 = #doc{
+        id = Id,
+        revs = {Pos, [RevId]},
+        deleted = Deleted,
+        body = Ptr
+    },
+    Doc1 = couch_db_engine:read_doc_body(SourceDb, Doc0),
+    #doc{body = Body, atts = AttInfos0} = Doc1,
+    External = if not is_binary(Body) -> ?term_size(body); true ->
```

Comment:
Surely you meant: `?term_size(Body)`

Reply:
Also see PR #1983 to improve this for the normal case.

Reply:
Doh! Good find re: ?term_size(body).  And will copy what we have in #1983

Reply:
Fixed in a fixup commit